lib/xgu/xgux.h

```diff
@@ -113,4 +113,72 @@ void xgux_set_transform_constant_matrix4x4(unsigned int location, unsigned int c
     }
 }
 
+#define xgux_vertex3f(x, y, z) \
+  xgux_set_vertex3f((x), (y), (z))
+#define xgux_vertex4f(x, y, z, w) \
+  xgux_set_vertex4f((x), (y), (z), (w))
+
+#if 0
+#define GENERIC_ATTRIBUTE_TYPE(suffix, type, name, register, extra_arguments, x, y, z, w) \
+inline \
+static void name ## suffix(extra_arguments, type, x, y, z, w) { \
+  xgux_set_vertex_data ## suffix(register, x, y, z, w);
+}
+
+#define GENERIC_ATTRIBUTE(name, register, extra_arguments, x, y, z, w) \
+  GENERIC_ATTRIBUTE_TYPE(1f, float, name, register, extra_arguments, x) \
+  GENERIC_ATTRIBUTE_TYPE(2f, float, name, register, extra_arguments, x, y) \
+  GENERIC_ATTRIBUTE_TYPE(3f, float, name, register, extra_arguments, x, y, z) \
+  GENERIC_ATTRIBUTE_TYPE(4f, float, name, register, extra_arguments, x, y, z, w)
+
+
+
+//FIXME: Also support XGU_POSITION_ARRAY?
+GENERIC_ATTRIBUTE(weight, XGU_WEIGHT_ARRAY,, x, y, z, w)
+GENERIC_ATTRIBUTE(normal, XGU_NORMAL_ARRAY,, x, y, z, w)
+GENERIC_ATTRIBUTE(diffuse, XGU_COLOR_ARRAY,, r, g, b, a)
+GENERIC_ATTRIBUTE(specular, XGU_COLOR_ARRAY,, r, g, b, a)
+GENERIC_ATTRIBUTE(fogcoord, XGU_POINT_SIZE_ARRAY,, x, y, z, w)
+GENERIC_ATTRIBUTE(point_size, XGU_POINT_SIZE_ARRAY,, x, y, z, w)
+GENERIC_ATTRIBUTE(back_diffuse, XGU_COLOR_ARRAY,, r, g, b, a)
+GENERIC_ATTRIBUTE(back_specular, XGU_COLOR_ARRAY,, r, g, b, a)
+GENERIC_ATTRIBUTE(texcoord, 8+index, unsigned int index, s, t, r, q)
+GENERIC_ATTRIBUTE(vertex_attribute, index, unsigned int index, x, y, z, w)
+
+#undef GENERIC_ATTRIBUTE
+#undef GENERIC_ATTRIBUTE_TYPE
+#endif
+
+//FIXME: Remove p argument?
+#define xgux_set_color3f(p, r, g, b) \
+  xgu_set_vertex_data4f(p, XGU_COLOR_ARRAY, (r), (g), (b), 1.0f)
+#define xgux_set_texcoord3f(index, s, t, r) \
+  xgu_set_vertex_data4f(p, 8+(index), (s), (t), (r), 1.0f)
+
+
+// This is as for D3D; you have to do this on your own:
+//
+//   size = viewport_height * size;
+//
+// This function then prepares it for this:
+//
+//   render_size = size * sqrt(1.0 / (constant_scale +
+//                                    linear_scale * distance +
+//                                    squared_scale * distance^2);
+//   render_size = clamp(render_size, minimum, maximum);
+//FIXME: Remove p argument?
+inline
+uint32_t* xgux_set_pointscale(uint32_t* p, float size, float constant_scale, float linear_scale, float squared_scale, float minimum, float maximum) {
```

Comment:
This should do what the specular / shininess functions do (suffix and treat this high level)